In [1]:
# 필요한 라이브러리 불러오기
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import koreanize_matplotlib
import seaborn as sns
import os
from ast import literal_eval  # 문자열로 저장된 딕셔너리를 진짜 딕셔너리로 변환

# GCS 파일 경로에서 데이터 불러오기 위한 라이브러리
from google.cloud import storage
from google.oauth2 import service_account

In [4]:

# 기본 정보 확인 함수 
def show_df_info(df, df_name):
    """
    DataFrame의 기본 정보와 통계 요약을 출력하는 함수
    """
    print(f"\n{df_name} 형태:")
    display(df.shape)
    print(f"\n{df_name} 첫 5행:")
    display(df.head())
    print(f"\n{df_name} 정보:")
    df.info()
    print(f"\n{df_name} 통계 요약:")
    display(df.describe())
    print(f"\n{df_name} 통계 요약:")
    display(df.describe(include = 'O'))
    print(f"\n{df_name} 결측치 확인:")
    display(df.isnull().sum())
    # print(f"\n{df_name} 중복값 확인:")
    # print(df.iloc[:,1:].duplicated().sum())

# 파일 불러오기 및 확인

In [69]:
# 230430 test
# GCS 파일 경로
gcs_path = "gs://part4_project/events/230430.parquet"
# 데이터 불러오기
test_df = pd.read_parquet(gcs_path, storage_options={"token": None}) # key_path // 오늘은 에러나서 None으로...

# 확인
display(test_df.head())
test_df.shape

event_datetime                    session_id  \
0  2023-05-01 09:00:00  jgGFnoyc6GWuGIEgjp3nUoKRQbF3   
1  2023-05-01 09:00:00  n3PI8GZnRFOKpJslIRgckMeRiln1   
2  2023-05-01 09:00:00  L31Dl8vRtdWPJ8XC0gGs1mPwzJH2   
3  2023-05-01 09:00:00  QjToElcYNkVxszJDHBhqtITlzDJ3   
4  2023-05-01 09:00:00  VRpXf303RGU730fIO4EjWstzAJi1   

                              device_id event_key description  \
0  cfc34eab-6930-488b-8109-70d37cec7dae    button          다음   
1  0dde2ecb-ff03-440f-98f3-94f20978b796    button     vote 선택   
2  b50774ed-24ce-49e1-9383-945ff3aea9ff    button        친구선택   
3  0999c6b6-c579-4c0f-916c-9a113010f84a    button      이름 셔플!   
4  45985493-ed35-4cc5-935f-1bcf922e213c    button     앱바 뒤로가기   

                  location devicemodel language  
0     signotherinputscreen  iPhone10,4    ko-KR  
1          homevotesscreen  iPhone14,4    ko-CA  
2        homeenquetescreen    SM-A125N       ko  
3        homeenquetescreen    SM-A305N       ko  
4  homefriendprofilescreen  iPhone14,5    ko-KR

(8588422, 8)

In [2]:
# 인증
key_path = "../config/codeit_project_vm_key.json"
credentials = service_account.Credentials.from_service_account_file(key_path)

# GCS 클라이언트 생성
client = storage.Client(credentials=credentials)
bucket = client.bucket("part4_project")

# 'events/' 경로에 있는 .csv 파일 리스트 가져오기
blobs = bucket.list_blobs(prefix="events/")
events_file_list = [f"gs://part4_project/{blob.name}" for blob in blobs if blob.name.endswith(".csv")] 

# 결과 확인
events_file_list

['gs://part4_project/events/df_230430.csv',
 'gs://part4_project/events/df_230506.csv',
 'gs://part4_project/events/df_230512.csv',
 'gs://part4_project/events/df_230518.csv',
 'gs://part4_project/events/df_230524.csv',
 'gs://part4_project/events/df_230530.csv',
 'gs://part4_project/events/df_230605.csv',
 'gs://part4_project/events/df_230611.csv',
 'gs://part4_project/events/df_230617.csv']

In [ ]:
# 파일들을 각각 변수로 불러오기
for title in events_file_list:
    # GCS 경로에서 파일명 추출
    file_name = title.split('/')[-1].replace('.csv', '')
    
    # Parquet 파일 읽기
    df = pd.read_csv(title, storage_options={"token": key_path})
    
    # 각각의 이름_df로 저장 (ex. device_properties_df)
    globals()[f"{file_name}"] = df
    
    # 출력 확인 (shape + head)
    print(f"\n {file_name} 불러오기 완료")
    print(f"→ shape: {df.shape}")
    print(f"→ 첫 5행:")
    display(df.head())
    
    # event_datetime이 있는 경우 날짜 변환 및 범위 출력
    if 'event_datetime' in df.columns:
        try:
            df['event_datetime'] = pd.to_datetime(df['event_datetime'])
            min_date = df['event_datetime'].min().date()
            max_date = df['event_datetime'].max().date()
            print(f"📆 {file_name}의 event_datetime 범위: {min_date} ~ {max_date}")
        except Exception as e:
            print(f"⚠️ {file_name}의 event_datetime 변환 중 오류 발생: {e}")
    else:
        print(f"⛔ {file_name}에는 'event_datetime' 컬럼이 존재하지 않음")


 df_230430 불러오기 완료
→ shape: (1200705, 8)
→ 첫 5행:


event_datetime                    session_id  \
0  2023-05-01 09:00:00  jgGFnoyc6GWuGIEgjp3nUoKRQbF3   
1  2023-05-01 09:00:00  n3PI8GZnRFOKpJslIRgckMeRiln1   
2  2023-05-01 09:00:00  L31Dl8vRtdWPJ8XC0gGs1mPwzJH2   
3  2023-05-01 09:00:00  QjToElcYNkVxszJDHBhqtITlzDJ3   
4  2023-05-01 09:00:00  VRpXf303RGU730fIO4EjWstzAJi1   

                              device_id event_key description  \
0  cfc34eab-6930-488b-8109-70d37cec7dae    button          다음   
1  0dde2ecb-ff03-440f-98f3-94f20978b796    button     vote 선택   
2  b50774ed-24ce-49e1-9383-945ff3aea9ff    button        친구선택   
3  0999c6b6-c579-4c0f-916c-9a113010f84a    button      이름 셔플!   
4  45985493-ed35-4cc5-935f-1bcf922e213c    button     앱바 뒤로가기   

                  location devicemodel language  
0     signotherinputscreen  iPhone10,4    ko-KR  
1          homevotesscreen  iPhone14,4    ko-CA  
2        homeenquetescreen    SM-A125N       ko  
3        homeenquetescreen    SM-A305N       ko  
4  homefriendprofilescreen  iPhone14,5    ko-KR

📆 df_230430_df의 event_datetime 범위: 2023-05-01 ~ 2023-05-01

 df_230506 불러오기 완료
→ shape: (1266122, 6)
→ 첫 5행:


event_datetime                            session_id  \
0  2023-05-07 09:00:00          OoWMTpTWpUa4m8I1RiKOk75Xnvs1   
1  2023-05-07 09:00:00          CrFi87nvwiVvHExxP9uSfIHOmBy1   
2  2023-05-07 09:00:00          W6frWgWJtbWTJwdm6eVWPGKwY2Y2   
3  2023-05-07 09:00:00  4EFB4D45-0AAF-424A-8E4E-2F19249C8892   
4  2023-05-07 09:00:00          JdxVFvtz6AcxxOZVvl6plHayjYJ3   

                              device_id       event_key devicemodel language  
0  498dbc0f-4537-41dc-93ec-f4406302a3b0    $session_end  iPhone14,5    ko-KR  
1  7770afa8-e228-4736-aa2a-658ebe2e3d25    $session_end  iPhone12,8    ko-KR  
2  9fbdc45d-a74a-4d9f-9d40-0f400b72877e    $session_end  iPhone14,8    ko-KR  
3  4efb4d45-0aaf-424a-8e4e-2f19249c8892  $session_start  iPhone12,3    ko-KR  
4  96348488-4d6a-49a8-8995-452718157fd7  $session_start    SM-A315N       ko

📆 df_230506_df의 event_datetime 범위: 2023-05-07 ~ 2023-05-09

 df_230512 불러오기 완료
→ shape: (1097522, 6)
→ 첫 5행:


Asia/Seoul                    session_id  \
0  2023-05-13T00:00:00.004Z  6zkImM4PD0MHIbJ9zRVXs6vX6e62   
1  2023-05-13T00:00:00.027Z  Sq5vui6fg2Nhz2EHeC8e4PWyxo13   
2  2023-05-13T00:00:00.027Z  RyWeiMNMACUhbawb63ITBMX1c1U2   
3  2023-05-13T00:00:00.046Z  3OaNLuqTVDTkMEc8IsZNBxF9hsH2   
4  2023-05-13T00:00:00.070Z  rWBKAsEaG3cpHuAPRyohFr3k6PB2   

                              device_id       event_key devicemodel language  
0  cb7a48f6-6882-4223-98fd-4fe73cb3903b  $session_start  iPhone14,4    ko-KR  
1  b2512274-1161-4fd2-a4e8-b40f1ecc0372    $session_end  iPhone12,8    ko-KR  
2  75f9bcd9-78c8-4dd6-a91e-0faf85a10526    $session_end  iPhone15,2    ko-KR  
3  380ebcac-4f9d-4004-968d-cf6372908adc  $session_start  iPhone12,1    en-KR  
4  290b1a00-659d-4d24-a26d-5524d7ac44ce    $session_end  iPhone14,3    ko-KR

⛔ df_230512_df에는 'event_datetime' 컬럼이 존재하지 않음

 df_230518 불러오기 완료
→ shape: (995594, 6)
→ 첫 5행:


event_datetime                            session_id  \
0  2023-05-19 09:00:00          nALWgF7izudSzKXArsiAOaYYsr23   
1  2023-05-19 09:00:00  E9A0131C-A70A-4917-88D8-257223F0A767   
2  2023-05-19 09:00:00          euhK4oLwdGPVT6sPNNN1Rs7xEKz1   
3  2023-05-19 09:00:00          TYAFQPAWozbqfQ3I296QuXJnWrx1   
4  2023-05-19 09:00:00          XhhuxAQvEcXdlD1JjbPC8H3AF602   

                              device_id       event_key devicemodel language  
0  c370e36d-b84c-4c97-af37-db356caf1741  $session_start  iPhone14,2    ko-KR  
1  e9a0131c-a70a-4917-88d8-257223f0a767    $session_end  iPhone11,2    ko-KR  
2  2ab9b1b1-41a7-4f6e-af31-85daac4509ea  $session_start  iPhone14,7    ko-KR  
3  c3236b64-a328-4743-8afd-461f85f1f1e0  $session_start  iPhone15,3    ko-KR  
4  efe3418d-ad59-4a20-a5f9-43d2195d1a08  $session_start  iPhone14,7    ko-KR

📆 df_230518_df의 event_datetime 범위: 2023-05-19 ~ 2023-05-20

 df_230524 불러오기 완료
→ shape: (903053, 6)
→ 첫 5행:


event_datetime                    session_id  \
0  2023-05-25 09:00:00  TeeVVRGrpuXE5q6f7pAu9ULVNvG3   
1  2023-05-25 09:00:00  9tcpOeHvyldCUXMgRyFVQI86rf13   
2  2023-05-25 09:00:00  aiDrtN8uHiQXRXmLMJhf27hKA4B3   
3  2023-05-25 09:00:00  Oqmt3cCrCQZAa0TT3B4s2kTOG1k1   
4  2023-05-25 09:00:00  GA1Nz2ehK2TmsRcqBzXQn6hkDHH2   

                              device_id       event_key devicemodel language  
0  b6afceaa-cc9c-4a66-b92f-9e202f9eb328  $session_start  iPhone14,7    ko-KR  
1  54a8af4d-47bf-4951-87c2-1817b26d1c31  $session_start  iPhone12,1    ko-KR  
2  b08ee1df-0bc6-4b38-9474-8cbc55fd6f17  $session_start  iPhone12,1    ko-KR  
3  ace73d0e-04d6-4d03-9460-a0a1143253f7  $session_start  iPhone11,8    ko-KR  
4  2adda5ce-1d37-43cd-8cbd-c75c06f84f74  $session_start  iPhone14,7    ko-KR

⚠️ df_230524_df의 event_datetime 변환 중 오류 발생: time data "2023-05" doesn't match format "%Y-%m-%d %H:%M:%S", at position 95147. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

 df_230530 불러오기 완료
→ shape: (2313, 6)
→ 첫 5행:


event_datetime                            session_id  \
0  2023-05-31 09:01:06          VAb6ai91iKPZZQhGKenhHZpAG5B2   
1  2023-05-31 09:02:20  fe556647-de19-43ff-9152-b3c38a8b4d26   
2  2023-05-31 09:02:46          LsAqqCTVtQPpevfSHXI8oRepVs73   
3  2023-05-31 09:03:50          NBfi1jm5FKfB6We7YfVRHFgeJY83   
4  2023-05-31 09:04:12  3CBFC179-1BB1-40A2-84EF-C39EFBD40C14   

                              device_id       event_key devicemodel language  
0  6DA6A8DC-1E61-4643-963A-ED3D782BE77A  $session_start  iPhone15,2    ko-KR  
1  fe556647-de19-43ff-9152-b3c38a8b4d26  $session_start    SM-G991N       ko  
2  5D47D549-EF22-41F3-A0D2-AD9EC1D21A2B  $session_start  iPhone15,2    ko-KR  
3  FD419DE8-B3CE-4630-B234-959ADA5F47CF  $session_start   iPhone9,3    ko-KR  
4  3CBFC179-1BB1-40A2-84EF-C39EFBD40C14  $session_start  iPhone14,7    ko-KR

📆 df_230530_df의 event_datetime 범위: 2023-05-31 ~ 2023-06-06

 df_230605 불러오기 완료
→ shape: (649, 5)
→ 첫 5행:


event_datetime                             device_id       event_key  \
0  2023-06-06 09:01:35  7918DF48-A239-45CE-AC1E-7E2E356454C4  $session_start   
1  2023-06-06 09:16:11  02F51978-FDD9-4C66-AE8D-820943E00AEF  $session_start   
2  2023-06-06 09:27:07  E75FFCB7-5AAB-4674-AE34-7A231136C367  $session_start   
3  2023-06-06 09:28:38  92D752F2-E7B9-4C01-892B-C9B3F1416B74  $session_start   
4  2023-06-06 09:44:06  47955d3d-b77f-47a6-9d3e-3fcb45a350df  $session_start   

  devicemodel language  
0  iPhone13,1    ko-KR  
1    iPad13,8    ko-KR  
2    iPad13,4    ko-KR  
3  iPhone15,2    ko-KR  
4    SM-A235N       ko

📆 df_230605_df의 event_datetime 범위: 2023-06-06 ~ 2023-06-12

 df_230611 불러오기 완료
→ shape: (263, 5)
→ 첫 5행:


event_datetime                             device_id       event_key  \
0  2023-06-12 09:00:33  8A5F41F9-D126-453C-8EA8-E0C2484584E3  $session_start   
1  2023-06-12 09:10:18  A81A2A45-6260-41C6-B81B-071F006D60BD  $session_start   
2  2023-06-12 09:17:21  CD41A81E-8D5C-4E7A-8289-587AD8C306F5  $session_start   
3  2023-06-12 09:21:00  5A92997F-362C-400B-8613-893A2FD6197F  $session_start   
4  2023-06-12 09:22:44  7FB46B64-B405-49C4-9378-8FABA73D4038  $session_start   

  devicemodel language  
0  iPhone13,1    ko-KR  
1  iPhone14,2    ko-KR  
2  iPhone13,1    ko-KR  
3  iPhone13,2    ko-KR  
4   iPad13,16    ko-KR

📆 df_230611_df의 event_datetime 범위: 2023-06-12 ~ 2023-06-18

 df_230617 불러오기 완료
→ shape: (1124983, 5)
→ 첫 5행:


event_datetime                             device_id       event_key  \
0  2023-06-18 09:07:08  8b6f9e35-473d-45d7-b8ad-884359609fdd  $session_start   
1  2023-06-18 09:07:11  8b6f9e35-473d-45d7-b8ad-884359609fdd    $session_end   
2  2023-06-18 09:07:34  4ff18855-0de6-4e51-a00c-7cdfe36bfc64  $session_start   
3  2023-06-18 10:01:12  99a6de2d-d222-46ff-81c2-729daff8e688  $session_start   
4  2023-06-18 11:30:47  4ff18855-0de6-4e51-a00c-7cdfe36bfc64  $session_start   

  devicemodel language  
0  iPhone15,2    ko-KR  
1  iPhone15,2    ko-KR  
2   iPhone8,4    ko-KR  
3    iPad13,1    ko-KR  
4   iPhone8,4    ko-KR

📆 df_230617_df의 event_datetime 범위: 2023-06-18 ~ 2023-06-23


In [ ]:
# 20230430의 event_datetime, session_id, device_id,	event_key, devicemodel,	language 컬럼만 가져오면 5월꺼랑 concat 가능
# 20230506 ~ 20230530 파일까지 concat 가능
# 20230605 ~ 20230617 파일까지 concat 가능

In [58]:
df_230524.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903053 entries, 0 to 903052
Data columns (total 6 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   event_datetime  903053 non-null  object
 1   session_id      903052 non-null  object
 2   device_id       903052 non-null  object
 3   event_key       903052 non-null  object
 4   devicemodel     903052 non-null  object
 5   language        903052 non-null  object
dtypes: object(6)
memory usage: 41.3+ MB


In [62]:
print(df_230524[df_230524.isnull().any(axis=1)])

df_230524 = df_230524.dropna() 

       event_datetime session_id device_id event_key devicemodel language
903052        2023-05        NaN       NaN       NaN         NaN      NaN


In [63]:
df_230524['event_datetime'] = pd.to_datetime(df_230524['event_datetime'])
df_230524['event_datetime'].min(), df_230524['event_datetime'].max()

(Timestamp('2023-05-25 09:00:00'), Timestamp('2023-05-26 15:04:21'))

# 230512 kst 로 변경

In [66]:
# 1. UTC 기준으로 파싱
df_230512['Asia/Seoul'] = pd.to_datetime(df_230512['Asia/Seoul'], utc=True)

# 2. 한국시간(Asia/Seoul)으로 타임존 변환
df_230512['Asia/Seoul'] = df_230512['Asia/Seoul'].dt.tz_convert('Asia/Seoul')
df_230512['Asia/Seoul'] = df_230512['Asia/Seoul'].dt.strftime('%Y-%m-%d %H:%M:%S')

# 3. 'Asia/Seoul' 컬럼명 event_datetime으로 변경
df_230512 = df_230512.rename(columns={'Asia/Seoul': 'event_datetime'})

print(df_230512.shape)
df_230512.head()

(1097522, 6)


event_datetime                    session_id  \
0  2023-05-13 09:00:00  6zkImM4PD0MHIbJ9zRVXs6vX6e62   
1  2023-05-13 09:00:00  Sq5vui6fg2Nhz2EHeC8e4PWyxo13   
2  2023-05-13 09:00:00  RyWeiMNMACUhbawb63ITBMX1c1U2   
3  2023-05-13 09:00:00  3OaNLuqTVDTkMEc8IsZNBxF9hsH2   
4  2023-05-13 09:00:00  rWBKAsEaG3cpHuAPRyohFr3k6PB2   

                              device_id       event_key devicemodel language  
0  cb7a48f6-6882-4223-98fd-4fe73cb3903b  $session_start  iPhone14,4    ko-KR  
1  b2512274-1161-4fd2-a4e8-b40f1ecc0372    $session_end  iPhone12,8    ko-KR  
2  75f9bcd9-78c8-4dd6-a91e-0faf85a10526    $session_end  iPhone15,2    ko-KR  
3  380ebcac-4f9d-4004-968d-cf6372908adc  $session_start  iPhone12,1    en-KR  
4  290b1a00-659d-4d24-a26d-5524d7ac44ce    $session_end  iPhone14,3    ko-KR

In [67]:
# # df_230512 한국시간 변경 버전 csv로 저장
# df_230512.to_csv('../data/df_230512_kst.csv', index=False)


# parquet으로 저장 
df_230512.to_parquet('../data/df_230512_kst.parquet', index=False)

In [68]:
# 230512 kst 버전 불러오기
df_230512_kst_test = pd.read_parquet("/home/codeit_project_vm/codeit_project/codeit-project-docker/data/df_230512_kst.parquet")

print(df_230512_kst_test.shape)
df_230512_kst_test.head()

(1097522, 6)


event_datetime                    session_id  \
0  2023-05-13 09:00:00  6zkImM4PD0MHIbJ9zRVXs6vX6e62   
1  2023-05-13 09:00:00  Sq5vui6fg2Nhz2EHeC8e4PWyxo13   
2  2023-05-13 09:00:00  RyWeiMNMACUhbawb63ITBMX1c1U2   
3  2023-05-13 09:00:00  3OaNLuqTVDTkMEc8IsZNBxF9hsH2   
4  2023-05-13 09:00:00  rWBKAsEaG3cpHuAPRyohFr3k6PB2   

                              device_id       event_key devicemodel language  
0  cb7a48f6-6882-4223-98fd-4fe73cb3903b  $session_start  iPhone14,4    ko-KR  
1  b2512274-1161-4fd2-a4e8-b40f1ecc0372    $session_end  iPhone12,8    ko-KR  
2  75f9bcd9-78c8-4dd6-a91e-0faf85a10526    $session_end  iPhone15,2    ko-KR  
3  380ebcac-4f9d-4004-968d-cf6372908adc  $session_start  iPhone12,1    en-KR  
4  290b1a00-659d-4d24-a26d-5524d7ac44ce    $session_end  iPhone14,3    ko-KR

In [54]:
df_230512_kst_test['event_datetime'] = pd.to_datetime(df_230512_kst_test['event_datetime'])
df_230512_kst_test.describe()

event_datetime
count                        1097522
mean   2023-05-13 20:38:04.599332608
min              2023-05-13 09:00:00
25%              2023-05-13 18:42:42
50%              2023-05-13 21:43:26
75%              2023-05-14 00:25:34
max              2023-05-14 03:01:39

# 2023.04.30 + 2023.05.06 ~ 2023.05.30. 파일 병합

In [6]:
# 20230430 파일의 event_datetime, session_id, device_id, event_key, devicemodel, language 컬럼만 가져와서 저장
df_230430_new = df_230430[['event_datetime', 'session_id', 'device_id', 'event_key', 'devicemodel', 'language']]

In [7]:
# 20230506 ~ 20230530 파일까지 concat

# 합치고 싶은 파일명 리스트
df_names = ['df_230506', 'df_230512', 'df_230518', 'df_230524', 'df_230530']

# globals()에서 해당 이름의 변수 가져와 리스트에 담기
dfs_to_concat = [globals()[name] for name in df_names]

# concat 병합
merged_0506_0530_df = pd.concat(dfs_to_concat, ignore_index=True)

# 결과 확인
print(f"병합된 DataFrame shape: {merged_0506_0530_df.shape}")
merged_0506_0530_df.head()

병합된 DataFrame shape: (4264604, 6)


event_datetime                            session_id  \
0  2023-05-07 09:00:00          OoWMTpTWpUa4m8I1RiKOk75Xnvs1   
1  2023-05-07 09:00:00          CrFi87nvwiVvHExxP9uSfIHOmBy1   
2  2023-05-07 09:00:00          W6frWgWJtbWTJwdm6eVWPGKwY2Y2   
3  2023-05-07 09:00:00  4EFB4D45-0AAF-424A-8E4E-2F19249C8892   
4  2023-05-07 09:00:00          JdxVFvtz6AcxxOZVvl6plHayjYJ3   

                              device_id       event_key devicemodel language  
0  498dbc0f-4537-41dc-93ec-f4406302a3b0    $session_end  iPhone14,5    ko-KR  
1  7770afa8-e228-4736-aa2a-658ebe2e3d25    $session_end  iPhone12,8    ko-KR  
2  9fbdc45d-a74a-4d9f-9d40-0f400b72877e    $session_end  iPhone14,8    ko-KR  
3  4efb4d45-0aaf-424a-8e4e-2f19249c8892  $session_start  iPhone12,3    ko-KR  
4  96348488-4d6a-49a8-8995-452718157fd7  $session_start    SM-A315N       ko

In [8]:
# 20230430과 5월 파일들 concat
merged_0430_0530_df = pd.concat([df_230430_new, merged_0506_0530_df], ignore_index=True)
merged_0430_0530_df.shape

(5465309, 6)

In [9]:
show_df_info(merged_0430_0530_df, "merged_0430_0530_df")


merged_0430_0530_df 형태:


(5465309, 6)


merged_0430_0530_df 첫 5행:


event_datetime                    session_id  \
0  2023-05-01 09:00:00  jgGFnoyc6GWuGIEgjp3nUoKRQbF3   
1  2023-05-01 09:00:00  n3PI8GZnRFOKpJslIRgckMeRiln1   
2  2023-05-01 09:00:00  L31Dl8vRtdWPJ8XC0gGs1mPwzJH2   
3  2023-05-01 09:00:00  QjToElcYNkVxszJDHBhqtITlzDJ3   
4  2023-05-01 09:00:00  VRpXf303RGU730fIO4EjWstzAJi1   

                              device_id event_key devicemodel language  
0  cfc34eab-6930-488b-8109-70d37cec7dae    button  iPhone10,4    ko-KR  
1  0dde2ecb-ff03-440f-98f3-94f20978b796    button  iPhone14,4    ko-CA  
2  b50774ed-24ce-49e1-9383-945ff3aea9ff    button    SM-A125N       ko  
3  0999c6b6-c579-4c0f-916c-9a113010f84a    button    SM-A305N       ko  
4  45985493-ed35-4cc5-935f-1bcf922e213c    button  iPhone14,5    ko-KR


merged_0430_0530_df 정보:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5465309 entries, 0 to 5465308
Data columns (total 6 columns):
 #   Column          Dtype 
---  ------          ----- 
 0   event_datetime  object
 1   session_id      object
 2   device_id       object
 3   event_key       object
 4   devicemodel     object
 5   language        object
dtypes: object(6)
memory usage: 250.2+ MB

merged_0430_0530_df 통계 요약:


event_datetime                    session_id  \
count               5465309                       5465308   
unique               405553                        696031   
top     2023-05-14 01:38:46  fT1BNDNQRyR5PBAgG9tg3dQkfnm1   
freq                    125                           985   

                                   device_id       event_key devicemodel  \
count                                5465305         5465304     5465304   
unique                                548955               3         635   
top     b66aeb3c-1e08-403a-8c73-14e02e3fb2c0  $session_start  iPhone14,5   
freq                                     985         2664451      681014   

       language  
count   5465304  
unique      203  
top       ko-KR  
freq    4123106


merged_0430_0530_df 통계 요약:


event_datetime                    session_id  \
count               5465309                       5465308   
unique               405553                        696031   
top     2023-05-14 01:38:46  fT1BNDNQRyR5PBAgG9tg3dQkfnm1   
freq                    125                           985   

                                   device_id       event_key devicemodel  \
count                                5465305         5465304     5465304   
unique                                548955               3         635   
top     b66aeb3c-1e08-403a-8c73-14e02e3fb2c0  $session_start  iPhone14,5   
freq                                     985         2664451      681014   

       language  
count   5465304  
unique      203  
top       ko-KR  
freq    4123106


merged_0430_0530_df 결측치 확인:


event_datetime    0
session_id        1
device_id         4
event_key         5
devicemodel       5
language          5
dtype: int64

In [10]:
# 결측치 확인 
merged_0430_0530_df[merged_0430_0530_df.isnull().any(axis = 1)]


event_datetime                    session_id  \
1200704  2023-05-01 23:27:41  qN1W72oUIfd4laYwk6Ux103cFX42   
2466826  2023-05-09 01:53:10                   BnHO3D8rBrM   
3564348  2023-05-14 03:01:39                    IYlwbLIgh4   
4559942  2023-05-20 02:23:13         4I9Exy4M1ogpOGgZNDPH5   
5462995              2023-05                           NaN   

                              device_id event_key devicemodel language  
1200704  f9ddd6f5-69d9-4639-81cd-ffd80a       NaN         NaN      NaN  
2466826                             NaN       NaN         NaN      NaN  
3564348                             NaN       NaN         NaN      NaN  
4559942                             NaN       NaN         NaN      NaN  
5462995                             NaN       NaN         NaN      NaN

In [11]:
# 저 2023-05는 230512 파일에서 온건가?? 
df_230512[df_230512['event_datetime'] == '2023-05']

Empty DataFrame
Columns: [event_datetime, session_id, device_id, event_key, devicemodel, language]
Index: []

- 결측치가 있는 4개의 행 중 event_datetime이 년-월 2023-05 까지만 있는 경우 존재
- 2023-05-12 파일에서 온 행은 아닌 것 같음...?

In [11]:
# 우선 결측치 있는 행 버리고 보자
merged_0430_0530_df = merged_0430_0530_df.dropna()
merged_0430_0530_df.shape

(5465304, 6)

In [12]:
# event_datetime 컬럼을 datetime 형식으로 변환
merged_0430_0530_df['event_datetime'] = pd.to_datetime(merged_0430_0530_df['event_datetime'])

In [13]:
# 기초 통계량 재확인
merged_0430_0530_df.describe(include = 'all')

event_datetime                    session_id  \
count                         5465304                       5465304   
unique                            NaN                        696028   
top                               NaN  fT1BNDNQRyR5PBAgG9tg3dQkfnm1   
freq                              NaN                           985   
mean    2023-05-12 23:59:18.904272128                           NaN   
min               2023-05-01 09:00:00                           NaN   
25%               2023-05-07 20:15:32                           NaN   
50%               2023-05-13 18:35:02                           NaN   
75%               2023-05-19 19:39:29                           NaN   
max               2023-06-06 08:55:03                           NaN   

                                   device_id       event_key devicemodel  \
count                                5465304         5465304     5465304   
unique                                548954               3         635   
top     b66aeb3c-1e08-403a-8c73-14e02e3fb2c0  $session_start  iPhone14,5   
freq                                     985         2664451      681014   
mean                                     NaN             NaN         NaN   
min                                      NaN             NaN         NaN   
25%                                      NaN             NaN         NaN   
50%                                      NaN             NaN         NaN   
75%                                      NaN             NaN         NaN   
max                                      NaN             NaN         NaN   

       language  
count   5465304  
unique      203  
top       ko-KR  
freq    4123106  
mean        NaN  
min         NaN  
25%         NaN  
50%         NaN  
75%         NaN  
max         NaN

- 2023-05-01 09:00:00 ~ 2023-06-06 08:55:03까지 있음(0430 ~ 0530 병합 시)
- session_id의 고유값 개수가 device_id 보다 많음
- event_key : session_start, session_end, button 3개인데, session_start가 가장 많음

In [ ]:
# event_key 고유값 개수 확인 
merged_0430_0530_df['event_key'].value_counts()  # button은 230430 파일만 있음 

event_key
$session_start    2664451
$session_end      1653814
button            1147039
Name: count, dtype: int64

- session_start가 session_end보다 많음 -> 모든게 짝지어지진 않을 확률 높음....
- button은 0430 파일에만 있는거라 사실 의미 없음.

In [16]:
# 무엇을 기준으로 봐야함??
# session_id 별로 device_id 몇 개인지
merged_0430_0530_df.groupby('session_id')['device_id'].nunique().reset_index().sort_values(by='device_id', ascending=False)

session_id  device_id
650373  uHVfeabKAfPVlsaVzCdbmhyByTC3         20
322619  ISz4K45Bu6RApABHjdl0DAh1Hn33         19
537785  gE1y248PkmTzwpzhKzWCMSzOjvg1         16
403259  Sa3OqUfCeLakNbD3G0qzarA10Pk2         16
13415   0JlVUGW5bdbW8nrM3halq8O2XtR2         16
...                              ...        ...
233869  BYoWnF1JpQXS2AtMBwgV7VMHipH2          1
233870  BYon1bp4WfPxLG3Xk9Gr0Tg32I32          1
233871  BYp9y9GPgXMowTasMma5a6cuIS32          1
233872  BYpap1fyARaNsUbAArZj96I0bEz1          1
348014  LeweIJqROjTitIVmTmhzyJ9o5IB3          1

[696028 rows x 2 columns]

In [ ]:
# 샘플 session_id 확인
merged_0430_0530_df[merged_0430_0530_df['session_id'] == 'uHVfeabKAfPVlsaVzCdbmhyByTC3']

event_datetime                    session_id  \
1212317 2023-05-07 09:18:28  uHVfeabKAfPVlsaVzCdbmhyByTC3   
1390493 2023-05-07 20:57:14  uHVfeabKAfPVlsaVzCdbmhyByTC3   
1409875 2023-05-07 21:27:27  uHVfeabKAfPVlsaVzCdbmhyByTC3   
1476544 2023-05-07 23:03:24  uHVfeabKAfPVlsaVzCdbmhyByTC3   
1566108 2023-05-08 01:12:30  uHVfeabKAfPVlsaVzCdbmhyByTC3   
1668301 2023-05-08 03:32:54  uHVfeabKAfPVlsaVzCdbmhyByTC3   
1802049 2023-05-08 06:16:17  uHVfeabKAfPVlsaVzCdbmhyByTC3   
1960919 2023-05-08 09:02:59  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2016428 2023-05-08 10:19:00  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2026262 2023-05-08 10:39:33  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2052001 2023-05-08 12:59:32  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2168733 2023-05-08 17:50:33  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2169463 2023-05-08 17:51:25  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2173029 2023-05-08 17:56:14  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2478059 2023-05-13 09:06:43  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2556554 2023-05-13 10:09:18  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2676298 2023-05-13 17:35:51  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2701496 2023-05-13 18:04:47  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2789973 2023-05-13 19:21:12  uHVfeabKAfPVlsaVzCdbmhyByTC3   
2933373 2023-05-13 20:57:18  uHVfeabKAfPVlsaVzCdbmhyByTC3   
3355831 2023-05-14 01:05:05  uHVfeabKAfPVlsaVzCdbmhyByTC3   
5307543 2023-05-26 07:44:06  uHVfeabKAfPVlsaVzCdbmhyByTC3   

                                    device_id       event_key devicemodel  \
1212317  56c268f4-dc82-4c70-8080-de52b2bc8c3e  $session_start  iPhone12,8   
1390493  3827cffd-c1f8-46ff-b078-2db64212b42d  $session_start  iPhone12,8   
1409875  a03d7d91-e820-4d3b-9ed0-4a1c60931539  $session_start  iPhone12,8   
1476544  79064dc0-f2f1-46c0-9dc0-b1e6eda23ac4  $session_start  iPhone12,8   
1566108  efccb5ec-b656-4019-a0c1-af597666d246  $session_start  iPhone12,8   
1668301  d0ccadb5-61ff-4109-b160-2ddb042c5f06  $session_start  iPhone12,8   
1802049  8e5d0723-2b2f-42d1-a38f-4290bc6aedfc  $session_start  iPhone12,8   
1960919  363cf2c2-e4f4-4aa0-992f-c4bf041f6d1a  $session_start  iPhone12,8   
2016428  2a2c65b4-aa09-4996-921c-e2cedaa33a14  $session_start  iPhone12,8   
2026262  b05008bc-b7bf-4a30-8e43-a570378b1aaf  $session_start  iPhone12,8   
2052001  67184043-dd68-4d81-8cb8-29ea3bee5a5d  $session_start  iPhone12,8   
2168733  66a9a11f-c527-4116-b05e-3eac949c706f  $session_start  iPhone12,8   
2169463  66a9a11f-c527-4116-b05e-3eac949c706f    $session_end  iPhone12,8   
2173029  66a9a11f-c527-4116-b05e-3eac949c706f  $session_start  iPhone12,8   
2478059  283504e2-f60a-40fc-af03-1407ee0dbc16  $session_start  iPhone12,8   
2556554  aef545d3-cf56-4a68-9a81-53b0b64b3f7f  $session_start  iPhone12,8   
2676298  b04d0e5d-6e1c-441d-9cb9-fc655e6696f6  $session_start  iPhone12,8   
2701496  bcd7d90a-de00-4c8c-b548-f5bb9a92d873  $session_start  iPhone12,8   
2789973  3d7999c9-8791-4177-acad-25244dcf0f2d  $session_start  iPhone12,8   
2933373  0dc2252a-50fc-47f5-b7c4-41fb36856799  $session_start  iPhone12,8   
3355831  a0d2b63d-bd4c-4b30-a736-6890169831a1  $session_start  iPhone12,8   
5307543  add34e51-3e94-4156-8159-de5101f4e9ae  $session_start  iPhone12,8   

        language  
1212317    ko-KR  
1390493    ko-KR  
1409875    ko-KR  
1476544    ko-KR  
1566108    ko-KR  
1668301    ko-KR  
1802049    ko-KR  
1960919    ko-KR  
2016428    ko-KR  
2026262    ko-KR  
2052001    ko-KR  
2168733    ko-KR  
2169463    ko-KR  
2173029    ko-KR  
2478059    ko-KR  
2556554    ko-KR  
2676298    ko-KR  
2701496    ko-KR  
2789973    ko-KR  
2933373    ko-KR  
3355831    ko-KR  
5307543    ko-KR

- device_id가 66a9a11f-c527-4116-b05e-3eac949c706f인 경우만 session_start 다음 session_end가 오지만 다시 session_start 로 마무리..
- 그 외 나머지는 모두 session_start 뿐임...

In [23]:
# device_id 별로 session_id 몇 개인지
merged_0430_0530_df.groupby('device_id')['session_id'].nunique().reset_index().sort_values(by='session_id', ascending=False)

device_id  session_id
372245  ad97ef42-477f-42c0-b1fd-3ec83098020a           5
54947   19a3d22c-5b44-49b6-93a5-d124d5245b4f           5
401879  bb47b31f-cf5b-41b6-b859-aafbd375b519           4
173037  50c7f22c-b6cd-4685-8af3-6f828f430f0b           4
270149  7e00cbd3-294f-4af5-9114-4ac58e466442           4
...                                      ...         ...
218101  65b91c67-77a8-4fd0-80ee-87d2230377ce           1
218100  65b90754-ddef-4c08-9aca-c3588e74e3dd           1
218098  65b8f4a2-7317-4e23-b1d2-9151d46a5d3b           1
218097  65b8eeb0-5017-4014-8d7c-6a0440150673           1
548953  ffff80ca-d517-40f4-bdef-e759411e1fe7           1

[548954 rows x 2 columns]

In [24]:
(merged_0430_0530_df.groupby('device_id')['session_id'].nunique().reset_index().sort_values(by='session_id', ascending=False)['session_id'] == 1).sum()

371262

- 548954 중에서 371262개는 session_id 1개
- 548954 중에서 177692개가 2개 이상의 session_id를 가지고 있음

In [14]:
# 샘플 device_id 확인
merged_0430_0530_df[merged_0430_0530_df['device_id'] == 'ad97ef42-477f-42c0-b1fd-3ec83098020a']

event_datetime                            session_id  \
270510  2023-05-01 10:17:52  AD97EF42-477F-42C0-B1FD-3EC83098020A   
270705  2023-05-01 10:17:58  AD97EF42-477F-42C0-B1FD-3EC83098020A   
270763  2023-05-01 10:17:59  AD97EF42-477F-42C0-B1FD-3EC83098020A   
270847  2023-05-01 10:18:01  AD97EF42-477F-42C0-B1FD-3EC83098020A   
270935  2023-05-01 10:18:03  AD97EF42-477F-42C0-B1FD-3EC83098020A   
...                     ...                                   ...   
281967  2023-05-01 10:22:33  AD97EF42-477F-42C0-B1FD-3EC83098020A   
2017473 2023-05-08 10:21:01  AD97EF42-477F-42C0-B1FD-3EC83098020A   
2017483 2023-05-08 10:21:02          fvscVUlvnnTqnH8FopkHiTIu3jg1   
2019039 2023-05-08 10:24:05          fvscVUlvnnTqnH8FopkHiTIu3jg1   
2019048 2023-05-08 10:24:06  AD97EF42-477F-42C0-B1FD-3EC83098020A   

                                    device_id       event_key devicemodel  \
270510   ad97ef42-477f-42c0-b1fd-3ec83098020a          button  iPhone13,2   
270705   ad97ef42-477f-42c0-b1fd-3ec83098020a          button  iPhone13,2   
270763   ad97ef42-477f-42c0-b1fd-3ec83098020a          button  iPhone13,2   
270847   ad97ef42-477f-42c0-b1fd-3ec83098020a          button  iPhone13,2   
270935   ad97ef42-477f-42c0-b1fd-3ec83098020a          button  iPhone13,2   
...                                       ...             ...         ...   
281967   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
2017473  ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
2017483  ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
2019039  ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
2019048  ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   

        language  
270510     ko-KR  
270705     ko-KR  
270763     ko-KR  
270847     ko-KR  
270935     ko-KR  
...          ...  
281967     ko-KR  
2017473    ko-KR  
2017483    ko-KR  
2019039    ko-KR  
2019048    ko-KR  

[115 rows x 6 columns]

In [ ]:
# event_key 가 button인 경우 제외하고 확인
merged_0430_0530_df[(merged_0430_0530_df['device_id'] == 'ad97ef42-477f-42c0-b1fd-3ec83098020a') & \
                     (merged_0430_0530_df['event_key'] != 'button')]

event_datetime                            session_id  \
271364  2023-05-01 10:18:15  AD97EF42-477F-42C0-B1FD-3EC83098020A   
271412  2023-05-01 10:18:16          jMa8EuiY23XL5yvOaiFL4Vcqxo32   
273083  2023-05-01 10:18:56          jMa8EuiY23XL5yvOaiFL4Vcqxo32   
273106  2023-05-01 10:18:57  AD97EF42-477F-42C0-B1FD-3EC83098020A   
273525  2023-05-01 10:19:08  AD97EF42-477F-42C0-B1FD-3EC83098020A   
273574  2023-05-01 10:19:10          jMa8EuiY23XL5yvOaiFL4Vcqxo32   
274000  2023-05-01 10:19:21          jMa8EuiY23XL5yvOaiFL4Vcqxo32   
274055  2023-05-01 10:19:22  AD97EF42-477F-42C0-B1FD-3EC83098020A   
275086  2023-05-01 10:19:49  AD97EF42-477F-42C0-B1FD-3EC83098020A   
275148  2023-05-01 10:19:51          YBj6GbybYJU17f2Z4YMAexV7ys72   
277057  2023-05-01 10:20:36          YBj6GbybYJU17f2Z4YMAexV7ys72   
277110  2023-05-01 10:20:37  AD97EF42-477F-42C0-B1FD-3EC83098020A   
280067  2023-05-01 10:21:46  AD97EF42-477F-42C0-B1FD-3EC83098020A   
280119  2023-05-01 10:21:48          YZi8n8VzB4S9EVp3JjWY9GFyTI63   
281923  2023-05-01 10:22:32          YZi8n8VzB4S9EVp3JjWY9GFyTI63   
281967  2023-05-01 10:22:33  AD97EF42-477F-42C0-B1FD-3EC83098020A   
2017473 2023-05-08 10:21:01  AD97EF42-477F-42C0-B1FD-3EC83098020A   
2017483 2023-05-08 10:21:02          fvscVUlvnnTqnH8FopkHiTIu3jg1   
2019039 2023-05-08 10:24:05          fvscVUlvnnTqnH8FopkHiTIu3jg1   
2019048 2023-05-08 10:24:06  AD97EF42-477F-42C0-B1FD-3EC83098020A   

                                    device_id       event_key devicemodel  \
271364   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
271412   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
273083   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
273106   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
273525   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
273574   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
274000   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
274055   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
275086   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
275148   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
277057   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
277110   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
280067   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
280119   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
281923   ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
281967   ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
2017473  ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
2017483  ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   
2019039  ad97ef42-477f-42c0-b1fd-3ec83098020a    $session_end  iPhone13,2   
2019048  ad97ef42-477f-42c0-b1fd-3ec83098020a  $session_start  iPhone13,2   

        language  
271364     ko-KR  
271412     ko-KR  
273083     ko-KR  
273106     ko-KR  
273525     ko-KR  
273574     ko-KR  
274000     ko-KR  
274055     ko-KR  
275086     ko-KR  
275148     ko-KR  
277057     ko-KR  
277110     ko-KR  
280067     ko-KR  
280119     ko-KR  
281923     ko-KR  
281967     ko-KR  
2017473    ko-KR  
2017483    ko-KR  
2019039    ko-KR  
2019048    ko-KR

- session_id가 같을 때 session_start와 session_end가 잘 짝지어져서 나옴
- 어떤 session_id는 device_id와 같은데 알파벳만 소문자 -> 대문자로 바뀜 (AD97EF42-477F-42C0-B1FD-3EC83098020A)

In [18]:
# devide_id 기준 하나 더 확인
# event_key 가 button인 경우 제외하고 확인
merged_0430_0530_df[(merged_0430_0530_df['device_id'] == 'bb47b31f-cf5b-41b6-b859-aafbd375b519') & \
                     (merged_0430_0530_df['event_key'] != 'button')]

event_datetime                            session_id  \
1319302 2023-05-07 18:37:07  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1319404 2023-05-07 18:37:25  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1319408 2023-05-07 18:37:26          bIuEKENvxKYdTHRTDj215UpDXAw2   
1326496 2023-05-07 18:56:04          bIuEKENvxKYdTHRTDj215UpDXAw2   
1347648 2023-05-07 19:41:47          bIuEKENvxKYdTHRTDj215UpDXAw2   
1348019 2023-05-07 19:42:36          bIuEKENvxKYdTHRTDj215UpDXAw2   
1387827 2023-05-07 20:52:59          bIuEKENvxKYdTHRTDj215UpDXAw2   
1388087 2023-05-07 20:53:23          bIuEKENvxKYdTHRTDj215UpDXAw2   
1388093 2023-05-07 20:53:23  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1388236 2023-05-07 20:53:38  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1388251 2023-05-07 20:53:39          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1580488 2023-05-08 01:33:09          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1638536 2023-05-08 02:53:26          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1640740 2023-05-08 02:56:25          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1640742 2023-05-08 02:56:25  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1641170 2023-05-08 02:57:02  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1641190 2023-05-08 02:57:03          bIuEKENvxKYdTHRTDj215UpDXAw2   
1641731 2023-05-08 02:57:43          bIuEKENvxKYdTHRTDj215UpDXAw2   
1641733 2023-05-08 02:57:43  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1641903 2023-05-08 02:57:57  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1641918 2023-05-08 02:57:58          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1663590 2023-05-08 03:26:34          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1819240 2023-05-08 06:35:17          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1820812 2023-05-08 06:36:59          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
1820815 2023-05-08 06:36:59  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1821627 2023-05-08 06:37:51  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1821642 2023-05-08 06:37:52          bIuEKENvxKYdTHRTDj215UpDXAw2   
1876276 2023-05-08 07:34:09          bIuEKENvxKYdTHRTDj215UpDXAw2   
1876278 2023-05-08 07:34:10  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1876565 2023-05-08 07:34:28  bb47b31f-cf5b-41b6-b859-aafbd375b519   
1876576 2023-05-08 07:34:29          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
2210044 2023-05-08 19:21:32          chdax9D5bWYKAxrtXeLPnfJ49Lp1   
2210046 2023-05-08 19:21:32  bb47b31f-cf5b-41b6-b859-aafbd375b519   
2210259 2023-05-08 19:22:02  bb47b31f-cf5b-41b6-b859-aafbd375b519   
2210271 2023-05-08 19:22:03          gKFzW5K0dqeOqKyxvUH9R8PWwbE2   
2210459 2023-05-08 19:22:28          gKFzW5K0dqeOqKyxvUH9R8PWwbE2   
2210460 2023-05-08 19:22:28  bb47b31f-cf5b-41b6-b859-aafbd375b519   
2211246 2023-05-08 19:24:35  bb47b31f-cf5b-41b6-b859-aafbd375b519   
2211252 2023-05-08 19:24:36          bIuEKENvxKYdTHRTDj215UpDXAw2   
2245691 2023-05-08 20:40:19          bIuEKENvxKYdTHRTDj215UpDXAw2   
2245692 2023-05-08 20:40:19  bb47b31f-cf5b-41b6-b859-aafbd375b519   
2245785 2023-05-08 20:40:31  bb47b31f-cf5b-41b6-b859-aafbd375b519   
2245801 2023-05-08 20:40:32          chdax9D5bWYKAxrtXeLPnfJ49Lp1   

                                    device_id       event_key devicemodel  \
1319302  bb47b31f-cf5b-41b6-b859-aafbd375b519  $session_start    SM-G981N   
1319404  bb47b31f-cf5b-41b6-b859-aafbd375b519    $session_end    SM-G981N   
1319408  bb47b31f-cf5b-41b6-b859-aafbd375b519  $session_start    SM-G981N   
1326496  bb47b31f-cf5b-41b6-b859-aafbd375b519    $session_end    SM-G981N   
1347648  bb47b31f-cf5b-41b6-b859-aafbd375b519  $session_start    SM-G981N   
1348019  bb47b31f-cf5b-41b6-b859-aafbd375b519    $session_end    SM-G981N   
1387827  bb47b31f-cf5b-41b6-b859-aafbd375b519  $session_start    SM-G981N   
1388087  bb47b31f-cf5b-41b6-b859-aafbd375b519    $session_end    SM-G981N   
1388093  bb47b31f-cf5b-41b6-b859-aafbd375b519  $session_start    SM-G981N   
1388236  bb47b31f-cf5b-41b6-b859-aafbd375b519    $session_end    SM-G981N   
1388251  bb47b31f-cf5b-41b6-b859-aafbd375b519  $session_start    SM-G981N   
1580488  bb47b31f-cf5b-41b6-b859-aafbd375b519    $se

- 이전에 확인한 device_id와 비슷한 양상을 보임
- session_id == device_id 인 경우가 있음 

In [20]:
# session_id와 device_id 같은 경우 확인
merged_0430_0530_df[merged_0430_0530_df['session_id'] == merged_0430_0530_df['device_id']]['session_id'].nunique()

67083

- session_id 고유값 개수 : 696028
- device_id 고유값 개수 : 548954     
그 중, 67083개가 동일함!! (대문자 <-> 소문자 까지 매칭한다면 더 많아질지도?)

In [36]:
# 대소문자를 무시하고 비교 (모두 소문자로 변환한 후 비교)
merged_0430_0530_df[merged_0430_0530_df['session_id'].str.lower() == merged_0430_0530_df['device_id'].str.lower()]['session_id'].str.lower().nunique()

199292

- 대소문자까지 변환해서 모두 같은 경우 199292개! 

In [30]:
# device_id 별로 session_id 개수 셌을 때, 1개인 경우는 device_id == session_id 인 경우일까? 
group_device_df = merged_0430_0530_df.groupby('device_id')['session_id'].nunique().reset_index(name = 'session_cnt')
# device_id에 session_id 1개인 경우의 device_id 추출
device_list = group_device_df[group_device_df['session_cnt'] == 1]['device_id'].unique()
device_list

array(['0000065b-8c82-4982-b67e-ac67b1e4c38e',
       '000007c9-e103-4eb5-9777-a9084d4952df',
       '00002245-458f-4cdd-8533-b448cd43dbd2', ...,
       'ffff2485-894d-4854-b4f0-a3b5c79c23d5',
       'ffff29f0-c716-42a8-9347-3efaf7c21ca6',
       'ffff80ca-d517-40f4-bdef-e759411e1fe7'], dtype=object)

In [33]:
merged_0430_0530_df[(merged_0430_0530_df['device_id'].isin(device_list)) & (merged_0430_0530_df['device_id'] == merged_0430_0530_df['session_id'])]

event_datetime                            session_id  \
23651   2023-05-01 09:05:16  dc8b8d7c-03b1-42f0-9541-a88c4f282e86   
23999   2023-05-01 09:05:20  dc8b8d7c-03b1-42f0-9541-a88c4f282e86   
24124   2023-05-01 09:05:21  dc8b8d7c-03b1-42f0-9541-a88c4f282e86   
24296   2023-05-01 09:05:23  dc8b8d7c-03b1-42f0-9541-a88c4f282e86   
25514   2023-05-01 09:05:38  dc8b8d7c-03b1-42f0-9541-a88c4f282e86   
...                     ...                                   ...   
5465291 2023-06-06 06:12:08  5ceb4e48-d1f9-4069-96b9-98e9957911af   
5465294 2023-06-06 06:26:30  7d6e0b00-d462-4ecb-a336-fa25adef3c61   
5465297 2023-06-06 07:18:25  B9CE934F-EDC8-43DC-A2DF-412EB4E05EB1   
5465304 2023-06-06 08:28:56  39275C5A-CEF4-4182-807B-19B2495AB6BF   
5465305 2023-06-06 08:46:22  9CCF2BAC-EAFA-45FD-AD19-61D656A4F0FF   

                                    device_id       event_key devicemodel  \
23651    dc8b8d7c-03b1-42f0-9541-a88c4f282e86          button    SM-G965N   
23999    dc8b8d7c-03b1-42f0-9541-a88c4f282e86          button    SM-G965N   
24124    dc8b8d7c-03b1-42f0-9541-a88c4f282e86          button    SM-G965N   
24296    dc8b8d7c-03b1-42f0-9541-a88c4f282e86          button    SM-G965N   
25514    dc8b8d7c-03b1-42f0-9541-a88c4f282e86          button    SM-G965N   
...                                       ...             ...         ...   
5465291  5ceb4e48-d1f9-4069-96b9-98e9957911af  $session_start    SM-N976N   
5465294  7d6e0b00-d462-4ecb-a336-fa25adef3c61  $session_start    SM-F711N   
5465297  B9CE934F-EDC8-43DC-A2DF-412EB4E05EB1  $session_start  iPhone13,2   
5465304  39275C5A-CEF4-4182-807B-19B2495AB6BF  $session_start  iPhone13,3   
5465305  9CCF2BAC-EAFA-45FD-AD19-61D656A4F0FF  $session_start  iPhone13,1   

        language  
23651         ko  
23999         ko  
24124         ko  
24296         ko  
25514         ko  
...          ...  
5465291       ko  
5465294       ko  
5465297    ko-KR  
5465304    ko-KR  
5465305    ko-KR  

[8483 rows x 6 columns]

- device_id에 session_id가 1개인 371262개 중에서 8483개만 device_id == session_id